src/pke/lib/scheme/bfvrns/bfvrns-leveledshe.cpp

```diff
@@ -412,4 +726,164 @@ void LeveledSHEBFVRNS::EvalMultCoreInPlace(Ciphertext<DCRTPoly>& ciphertext, con
+    if (index == 0) {
+        Ciphertext<DCRTPoly> result = ciphertext->Clone();
+        return result;
+    }
+
+    const auto cc = ciphertext->GetCryptoContext();
+
+    usint autoIndex = FindAutomorphismIndex(index, m);
+
+    auto evalKey = cc->GetEvalAutomorphismKeyMap(ciphertext->GetKeyTag()).find(autoIndex)->second;
+
+    auto algo                       = cc->GetScheme();
+    const std::vector<DCRTPoly>& cv = ciphertext->GetElements();
+
+    std::shared_ptr<std::vector<DCRTPoly>> ba = algo->EvalFastKeySwitchCore(digits, evalKey, cv[0].GetParams());
+
+    const auto cryptoParams = std::dynamic_pointer_cast<CryptoParametersBFVRNS>(ciphertext->GetCryptoParameters());
+
+    if (cryptoParams->GetMultiplicationTechnique() == HPSPOVERQLEVELED) {
+        size_t levels   = ciphertext->GetDepth() - 1;
+        size_t sizeQ    = cv[0].GetNumOfElements();
+        double dcrtBits = cv[0].GetElementAtIndex(0).GetModulus().GetMSB();
+        // how many levels to drop
+        uint32_t levelsDropped = FindLevelsToDrop(levels, cryptoParams, dcrtBits, false);
+        uint32_t l             = levelsDropped > 0 ? sizeQ - 1 - levelsDropped : sizeQ - 1;
+
+        (*ba)[0].ExpandCRTBasisQlHat(cryptoParams->GetElementParams(), cryptoParams->GetQlHatModq(l),
+                                     cryptoParams->GetQlHatModqPrecon(l), sizeQ);
+        (*ba)[1].ExpandCRTBasisQlHat(cryptoParams->GetElementParams(), cryptoParams->GetQlHatModq(l),
+                                     cryptoParams->GetQlHatModqPrecon(l), sizeQ);
+    }
+
+    usint N = cryptoParams->GetElementParams()->GetRingDimension();
+    std::vector<usint> vec(N);
+    PrecomputeAutoMap(N, autoIndex, &vec);
+
+    (*ba)[0] += cv[0];
+
+    (*ba)[0] = (*ba)[0].AutomorphismTransform(autoIndex, vec);
+    (*ba)[1] = (*ba)[1].AutomorphismTransform(autoIndex, vec);
+
+    Ciphertext<DCRTPoly> result = ciphertext->Clone();
+
+    result->SetElements({std::move((*ba)[0]), std::move((*ba)[1])});
+
+    return result;
+}
+
+usint LeveledSHEBFVRNS::FindAutomorphismIndex(usint index, usint m) const {
+    return FindAutomorphismIndex2n(index, m);
+}
+
+void LeveledSHEBFVRNS::RelinearizeCore(Ciphertext<DCRTPoly>& ciphertext, const EvalKey<DCRTPoly> evalKey) const {
+    const auto cryptoParams = std::dynamic_pointer_cast<CryptoParametersBFVRNS>(ciphertext->GetCryptoParameters());
+    uint32_t l              = 0;
+
+    std::vector<DCRTPoly>& cv = ciphertext->GetElements();
+
+    auto algo = ciphertext->GetCryptoContext()->GetScheme();
+
+    if (cryptoParams->GetMultiplicationTechnique() == HPSPOVERQLEVELED) {
+        size_t levels   = ciphertext->GetDepth() - 1;
+        size_t sizeQ    = cv[0].GetNumOfElements();
+        double dcrtBits = cv[0].GetElementAtIndex(0).GetModulus().GetMSB();
+
+        // how many levels to drop
+        uint32_t levelsDropped = FindLevelsToDrop(levels, cryptoParams, dcrtBits, false);
```

Comment:
Why is the `keySwitch` flag set to false? This applies to other calls to FindLevelsToDrop as well.

Reply:
fixed